### DISCUSSION TITLE
Dudas con la 4ta iteracion de QMP

### DISCUSSION THREAD

**Discussion by LeoCubelli** [dds-jv/foro#232] [Consultas de objetos] [answered]
Buenas, según tengo entendido un atuendo es un conjunto de prendas ya definido. Por ejemplo: Atuendo atuendo1 = new Atuendo(Prendas), en este caso dentro de prendas puedo pasar tres gorras, como podría validar que un atuendo sea correcto? hay una estructura de 1 o mas accesorios, 1 parte superior, 1 parte inferior, 1 calzado?
En la 4ta iteración se agrega la temperatura adecuada para cada prenda, y el atuendo que se genera es de una prenda random que cumpla con la temperatura de cada categoría (parte superior, inferior, calzado)?
Hasta ahora tengo como Atuendos predefinidos, adentro de un "Ropero" y a Atuendo le agregué un atributo de la clase Temperatura que tiene un máximo y un mínimo, para que sugiera los atuendos predefinidos. Pero los requerimientos hablan de tener la temperatura en cada prenda y no se si esto estaría bien.

**Comment by flbulgarelli** [comment, marked as answer]
> Buenas, según tengo entendido un atuendo es un conjunto de prendas ya definido. Por ejemplo: Atuendo atuendo1 = new Atuendo(Prendas), en este caso dentro de prendas puedo pasar tres gorras, como podría validar que un atuendo sea correcto? hay una estructura de 1 o mas accesorios, 1 parte superior, 1 parte inferior, 1 calzado?

Sí, más que pensar a un atuendo como una colección genérica de prendas, te va a convenir definir una estructura más o menos rígida como la que describís: 

1. 0..N accesorios
2. 1..N parte superior
3. 1 parte inferior
4. 1 calzado

De hecho, para empezar podrías plantear que la parte suprior y los accesorios son (exactamente o como máximo, respectivamente) 1 en lugar de N, y luego cuando tengas todo más o menos encarado, generalizar ese aspecto. 

> En la 4ta iteración se agrega la temperatura adecuada para cada prenda, y el atuendo que se genera es de una prenda random que cumpla con la temperatura de cada categoría (parte superior, inferior, calzado)?

Correcto. En la parte superior, deberías considerar la aditividad del nivel de abrigo. 

> Hasta ahora tengo como Atuendos predefinidos, adentro de un "Ropero" 

Ojo, en el ropero vos deberías tener, justamente, ropa, no atuendos (porque los atuendos los construís a partir de la ropa en el ropero, aplicando combinatorias). 

> a Atuendo le agregué un atributo de la clase Temperatura que tiene un máximo y un mínimo, para que sugiera los atuendos predefinidos. Pero los requerimientos hablan de tener la temperatura en cada prenda y no se si esto estaría bien.

No, no sería correcto. ¿No podrías calcular acaso el nivel de abrigo de un atuendo a partir del nivel de abrigo de las ropas que lo componen?

**Reply by LeoCubelli** [reply]
> > En la 4ta iteración se agrega la temperatura adecuada para cada prenda, y el atuendo que se genera es de una prenda random que cumpla con la temperatura de cada categoría (parte superior, inferior, calzado)?
> 
> Correcto. En la parte superior, deberías considerar la aditividad del nivel de abrigo.
> 
Como sería eso de la aditividad? Cuando sugiere una parte superior, ésta tiene que ser apta para la temperatura actual, por ejemplo: si en el ropero tengo 5 buzos (con una temp maxima de 20 grados, osea que cuando pasa los 20 grados no te la tendria que recomendar), 3 camperas (temp max de 10 grados), 10 remeras (temp max 40 grados), si un día hace 5 grados y me pide una sugerencia, le tendría que devolver todas las prendas ya que todas son recomendadas para 5 grados, o te puede elegir solo una de esas prendas (puede ser solo una remera en un día que hace 5 grados). No se si me explico

> > Hasta ahora tengo como Atuendos predefinidos, adentro de un "Ropero"
> 
> Ojo, en el ropero vos deberías tener, justamente, ropa, no atuendos (porque los atuendos los construís a partir de la ropa en el ropero, aplicando combinatorias).
> 
Y esas combinatorias quien las define?

**Reply by flbulgarelli** [reply]
> Y esas combinatorias quien las define?

Es una responsabilidad del sistema. Justamente eso lo tenés que definir vos como parte de tu algoritmo de combinatorias. El mismo se desarrolla en [QMP3](https://docs.google.com/document/d/1th_F_BsF91zKuFWk6POogegljhrVW9ThybW_9U_xJZY/edit#heading=h.uyku9mnteh0t) (que es bonus). Acá hay [una solución posible](https://docs.google.com/document/d/1eMm7sRyyi-PJ8YFt0Ro2ecxfHBC6WJBlui7x21rFA2Q/edit#). Si no querés adentrarte en ese problema, podés simplemente asumir que ya existe un componente que lo resuelve. 

> No se si me explico

No estoy seguro de haber entendido del todo tu ejemplo, pero si hace una temperatura tal que con solo una prenda podés cubrir sus necesidades de abrigo, esa prenda por sí misma puede ser la parte superior de una sugerencia. Y por otro lado, si tenés una temperatura (por ejemplo -20 grados respecto de tu temperatura "base") para la cual no hay una prenda que la cubra, pero sí dos o más que sumadas lo logran (por ejemplo, una remera que aporta 2 grados + un buzo que aporta 8  + una campera que aporta 10), entonces deberías sugerir la combinación de las tres prendas. 

Acá como ves estoy asumiendo varias cosas (por ejemplo que existe una temperatura "base" que quizás se ajusta al inicio de la aplicación) o que cualquier combinación de prendas superiores es válida (lo cual en realidad es falso, porque no podés ponerte  tres camperas, por ejemplo), pero es un punto de partida para llegar a una solución parcial.  

Nuevamente, el problema es complejo, por eso no pedimos QMP3 como obligatorio.

**Reply by LeoCubelli** [reply]
Perfecto, no sabia que esa parte estaba en la 3ra iteración. Ahora con esto se me vino otra duda, la temperatura se define en la prenda? Porque también se podría definir en los TiposDePrenda, toda REMERA cubre 2 grados, toda CAMISA 3 grados, pero supongo que depende de como lo quiera implementar.

**Reply by flbulgarelli** [reply]
> pero supongo que depende de como lo quiera implementar.

Exacto, para QMP4 da lo mismo. 

Pero si hilamos más fino en realidad una decisión u otra va más allá del diseño: te cambia la funcionalidad. Ponerlo en el tipo de prenda hace a las cosas sencillas (todos los buzo abrigan siempre lo mismo, todas las camperas también, y así) pero perdés la posibilidad de que yo en tanto usuario pueda entrar a la aplicación y cargar a un buzo con un cierto "nivel de abrigo", y distinguir aún dentro de la misma categoría aquellas prendas que son más livianas o más gruesas.
